auto-value-gson/src/main/java/com/ryanharter/auto/value/gson/AutoValueGsonExtension.java

```diff
@@ -587,6 +519,120 @@ public MethodSpec createReadMethod(ClassName className,
     return readMethod.build();
   }
 
+  private MethodSpec createReadMethod(ClassName className,
+      TypeName autoValueClassName,
+      List<Property> properties,
+      ImmutableMap<Property, FieldSpec> adapters,
+      TypeElement autoValueClassElement,
+      BuilderContext builderContext,
+      ProcessingEnvironment processingEnvironment) {
+    ParameterSpec jsonReader = ParameterSpec.builder(JsonReader.class, "jsonReader").build();
+    MethodSpec.Builder readMethod = MethodSpec.methodBuilder("read")
+        .addAnnotation(Override.class)
+        .addModifiers(PUBLIC)
+        .returns(autoValueClassName)
+        .addParameter(jsonReader)
+        .addException(IOException.class);
+
+    ClassName token = ClassName.get(JsonToken.NULL.getDeclaringClass());
+
+    readMethod.beginControlFlow("if ($N.peek() == $T.NULL)", jsonReader, token);
+    readMethod.addStatement("$N.nextNull()", jsonReader);
+    readMethod.addStatement("return null");
+    readMethod.endControlFlow();
+
+    Optional<ExecutableElement> defaultBuilder = findDefaultBuilder(autoValueClassElement, builderContext.builderClass());
+
+    ClassName builderClassName = ClassName.get(builderContext.builderClass());
+    FieldSpec builder = FieldSpec.builder(builderClassName, "_builder").build();
+    if (defaultBuilder.isPresent()) {
+      ClassName baseClass = ClassName.get(autoValueClassElement);
+      readMethod.addStatement("$T $N = $T.$L", builder.type, builder, baseClass, defaultBuilder.get());
+    } else {
+      ClassName finalBuilderClassName = className.nestedClass(builderClassName.simpleName());
+      readMethod.addStatement("$T $N = new $T()", builder.type, builder, finalBuilderClassName);
+    }
+
+    readMethod.addStatement("$N.beginObject()", jsonReader);
+
+    readMethod.beginControlFlow("while ($N.hasNext())", jsonReader);
+
+    FieldSpec name = FieldSpec.builder(String.class, "_name").build();
+    readMethod.addStatement("$T $N = $N.nextName()", name.type, name, jsonReader);
+
+    readMethod.beginControlFlow("if ($N.peek() == $T.NULL)", jsonReader, token);
+    readMethod.addStatement("$N.nextNull()", jsonReader);
+    readMethod.addStatement("continue");
+    readMethod.endControlFlow();
+
+    readMethod.beginControlFlow("switch ($N)", name);
+    for (Property prop : properties) {
+      if (!prop.shouldDeserialize()) {
+        continue;
+      }
+
+      for (String alternate : prop.serializedNameAlternate()) {
+        readMethod.addCode("case $S:\n", alternate);
+      }
+      readMethod.beginControlFlow("case $S:", prop.serializedName());
+      Set<ExecutableElement> setters = builderContext.setters().get(prop.humanName);
+
+      // search overloaded methods for a type match
+      Optional<CodeBlock> setterBlock = setters.stream()
+          .filter(e -> ClassName.get(e.getParameters().get(0).asType()).equals(prop.type))
```

Comment:
It should be ok because `builderContext.setters()` is a multimap of properties to their setter(s).